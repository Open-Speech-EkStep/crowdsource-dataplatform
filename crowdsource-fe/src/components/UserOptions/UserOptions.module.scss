.root:global(.show) {
  position: relative;

  @include media-breakpoint-up(lg) {
    &::before {
      background-color: var(--primary-color);
      bottom: 0;
      content: '';
      height: 0.125rem;
      position: absolute;
      width: 100%;
    }
  }

  .toggle {
    background-color: var(--text-white);
    border-radius: 0;

    @include media-breakpoint-up(lg) {
      background-color: rgba(159, 200, 255, 0.39);
    }
  }
}

.toggle {
  @include media-breakpoint-up(md) {
    width: 4.5rem;
  }

  @include media-breakpoint-up(xl) {
    width: auto;
  }

  &:hover {
    background-color: var(--text-white);
    color: var(--primary-color);
  }
}

.toggle,
.toggle:focus {
  background-color: var(--text-white);
  border: 0;
  border-radius: 0;
}

.item {
  color: var(--primary-color);
  font-size: 0.875rem;
  padding: 0.5rem 1rem;

  @include media-breakpoint-up(md) {
    font-size: 1rem;
  }

  &:visited,
  &:hover,
  &:active {
    background-color: rgba(159, 200, 255, 0.39);
    color: var(--primary-color);
  }
}
