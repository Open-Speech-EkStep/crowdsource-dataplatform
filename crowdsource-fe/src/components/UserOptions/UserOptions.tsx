import { Fragment, useState } from 'react';

import classnames from 'classnames';
import { useTranslation } from 'next-i18next';
import dynamic from 'next/dynamic';
import Image from 'next/image';
import Dropdown from 'react-bootstrap/Dropdown';

import Link from 'components/Link';
import localStorageConstants from 'constants/localStorageConstants';
import routePaths from 'constants/routePaths';
import useLocalStorage from 'hooks/useLocalStorage';
import type SpeakerDetails from 'types/SpeakerDetails';

import styles from './UserOptions.module.scss';

const ChangeUserModal = dynamic(() => import('components/ChangeUserModal'), { ssr: false });

const UserOptions = () => {
  const { t } = useTranslation();
  const [modalShow, setModalShow] = useState(false);
  const [speakerDetails] = useLocalStorage<SpeakerDetails>(localStorageConstants.speakerDetails);

  if (!speakerDetails) {
    return null;
  }

  return (
    <Fragment>
      <Dropdown
        data-testid="UserOptions"
        id="userOptions"
        className={`${styles.root} position-relative`}
        align="end"
      >
        <Dropdown.Toggle
          id="languageSwitcherToggle"
          variant="light"
          className={classnames(styles.toggle, 'd-flex h-100 justify-content-center align-items-center')}
        >
          <Image src="/images/usericon.svg" width="24" height="24" alt={t('languageIconAlt')} />
          {speakerDetails.userName && (
            <span className="d-none d-xl-block mx-1">{speakerDetails.userName}</span>
          )}
        </Dropdown.Toggle>

        <Dropdown.Menu>
          {!speakerDetails.userName && (
            <Dropdown.Item className={classnames(styles.item, 'disabled')}>({t('noUsername')})</Dropdown.Item>
          )}
          <Dropdown.Item
            className={`${styles.item} text-primary display-5 py-2 px-4`}
            onClick={() => setModalShow(true)}
          >
            {t('changeUser')}
          </Dropdown.Item>
          <Link href={routePaths.myBadges} passHref>
            <Dropdown.Item className={`${styles.item} text-primary display-5 py-2 px-4`}>
              {t('myBadges')}
            </Dropdown.Item>
          </Link>
        </Dropdown.Menu>
      </Dropdown>
      <ChangeUserModal show={modalShow} onHide={() => setModalShow(false)} />
    </Fragment>
  );
};

export default UserOptions;
