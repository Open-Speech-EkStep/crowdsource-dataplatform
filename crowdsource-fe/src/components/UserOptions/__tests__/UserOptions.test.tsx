import { when } from 'jest-when';
import router from 'next/router';

import {
  render,
  verifyAxeTest,
  screen,
  userEvent,
  waitFor,
  waitForElementToBeRemoved,
} from 'utils/testUtils';

import UserOptions from '../UserOptions';

describe('UserOptions', () => {
  const setup = () => {
    when(localStorage.getItem)
      .calledWith('speakerDetails')
      .mockImplementation(() => null);

    return render(<UserOptions />);
  };
  const setupWithSpeakerDetails = (data = {}) => {
    const speakerDetails = {
      gender: 'male',
      age: '',
      motherTongue: 'Hindi',
      userName: 'abcd',
      language: 'Hindi',
      toLanguage: '',
      ...data,
    };
    when(localStorage.getItem)
      .calledWith('speakerDetails')
      .mockImplementation(() => JSON.stringify(speakerDetails));

    return render(<UserOptions />);
  };

  verifyAxeTest(() => setup());
  verifyAxeTest(() => setupWithSpeakerDetails());

  it('should not render the component when speaker details are not present', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the component when speaker details are present', () => {
    const { asFragment } = setupWithSpeakerDetails();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the component when speaker details are present but username is not present', () => {
    const { asFragment } = setupWithSpeakerDetails({
      userName: '',
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should hide ChangeUserModal initially', () => {
    setupWithSpeakerDetails();

    expect(screen.queryByTestId('ChangeUserModal')).not.toBeInTheDocument();
  });

  it('should show ChangeUserModal when change user option is clicked', async () => {
    setupWithSpeakerDetails();

    userEvent.click(screen.getByRole('button', { name: 'languageIconAlt abcd' }));

    await waitFor(() => expect(screen.getByRole('button', { name: 'changeUser' })).toBeInTheDocument());

    userEvent.click(screen.getByRole('button', { name: 'changeUser' }));

    expect(screen.getByTestId('ChangeUserModal')).toBeInTheDocument();

    userEvent.click(screen.getByRole('button', { name: 'Close' }));

    await waitForElementToBeRemoved(() => screen.queryByTestId('ChangeUserModal'));

    expect(screen.queryByTestId('ChangeUserModal')).not.toBeInTheDocument();
  });

  it('should navigate to my badges page on "my badges" link click', async () => {
    setupWithSpeakerDetails();

    userEvent.click(screen.getByRole('button', { name: 'languageIconAlt abcd' }));

    await waitFor(() => expect(screen.getByRole('button', { name: 'changeUser' })).toBeInTheDocument());

    expect(router.pathname).toBe('');

    userEvent.click(screen.getByRole('link', { name: 'myBadges' }));

    expect(router.pathname).toBe('/my-badges.html');
  });
});
