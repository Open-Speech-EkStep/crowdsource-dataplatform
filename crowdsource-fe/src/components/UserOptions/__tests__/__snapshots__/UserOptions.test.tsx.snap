// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserOptions should not render the component when speaker details are not present 1`] = `<DocumentFragment />`;

exports[`UserOptions should render the component when speaker details are present 1`] = `
<DocumentFragment>
  <div
    class="root position-relative dropdown"
    data-testid="UserOptions"
    id="userOptions"
  >
    <button
      aria-expanded="false"
      class="toggle d-flex h-100 justify-content-center align-items-center px-3 dropdown-toggle btn btn-light"
      id="userOptions"
      type="button"
    >
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <img
          alt="Language Icon"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </div>
      <span
        class="d-none d-xl-block mx-1"
      >
        abcd
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`UserOptions should render the component when speaker details are present but username is not present 1`] = `
<DocumentFragment>
  <div
    class="root position-relative dropdown"
    data-testid="UserOptions"
    id="userOptions"
  >
    <button
      aria-expanded="false"
      class="toggle d-flex h-100 justify-content-center align-items-center px-3 dropdown-toggle btn btn-light"
      id="userOptions"
      type="button"
    >
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <img
          alt="Language Icon"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </div>
    </button>
  </div>
</DocumentFragment>
`;
